src/concurrent.rs

```diff
@@ -289,33 +300,58 @@ mod tests {
         assert_eq!(0, reader2.read(&mut buffer).expect("Expected 0 byte count as EOF after close"));
     }
 
-    #[test]
-    fn test_shareable_file_close_unblocks_reads_without_error() {
+    fn try_shareable_file_close_unblocks_reads_without_error() {
         let dir = tempfile::tempdir().unwrap();
         let path = dir.path().join("pipe");
         unistd::mkfifo(&path, sys::stat::Mode::S_IRUSR | sys::stat::Mode::S_IWUSR).unwrap();
 
-        // We need to write to the FIFO for the read-only open below to not block.
+        // We need to open the FIFO for writes so that the open for reads doesn't block.  But we
+        // have to do this open in a separate thread because it will block too.
+        let (writer_unblock_tx, writer_unblock_rx) = mpsc::channel();
         let writer_handle = {
             let path = path.clone();
             thread::spawn(move || {
-                let mut writer = fs::File::create(&path).unwrap();
-                write!(writer, "some text longer than we'll ever read")
+                let _keep_open = fs::File::create(path).unwrap();
+                // Keep the file open until the writer is unblocked.
+                writer_unblock_rx.recv()
             })
         };
 
+        // This will block until the thread we spawned above opens the file for writing.
         let mut file = ShareableFile::from(fs::File::open(&path).unwrap());
 
+        let (reader_ready_tx, reader_ready_rx) = mpsc::channel();
         let reader_handle = {
             let mut file = file.clone_unowned();
             thread::spawn(move || {
                 let mut buffer = [0];
+                reader_ready_tx.send(()).unwrap();
                 file.read(&mut buffer)
             })
         };
 
-        writer_handle.join().unwrap().expect("Write didn't finish successfully");
+        // This is racy: there is no guarantee that the reader thread is now blocked on the read
+        // syscall although there is a high likelihood that it is.  If we experience the race, the
+        // test will silently pass even if there is a problem in the ShareableFile.
+        reader_ready_rx.recv().unwrap();
```

Comment:
Does it make sense to add a small sleep to raise the probability of blocking before this line is triggered? I can imagine that if you are running the tests on a single core machine the scheduler could predictably prefer one thread over the other. Even a small sleep should make that unlikely to always occur on any machine.

Reply:
Yes. Done.

This had crossed my mind but for some reason I was thinking "I have to sleep at least one second, and that'd be too annoying on unit tests". But 1ms is fine.

Reply:
You want the sleep after the receive.

Reply:
Oops, done.